public/js/controllers/attackcost_controller.js

```diff
@@ -0,0 +1,385 @@
+import { Controller } from 'stimulus'
+import TurboQuery from '../helpers/turbolinks_helper'
+import { getDefault } from '../helpers/module_helper'
+import globalEventBus from '../services/event_bus_service'
+import { nightModeOptions } from '../helpers/chart_helper'
+import dompurify from 'dompurify'
+
+function digitformat (amount, decimalPlaces, noComma) {
+  if (!amount) return 0
+
+  if (noComma) return amount.toFixed(decimalPlaces)
+
+  decimalPlaces = decimalPlaces || 0
+  let result = parseFloat(amount).toLocaleString(undefined, { minimumFractionDigits: decimalPlaces, maximumFractionDigits: decimalPlaces }).replace(/\.0*$/, '')
+  if (result.indexOf('.') > -1 && result.endsWith('0')) {
+    return removeTrailingZeros(result)
+  }
+
+  return result
+}
+
+function removeTrailingZeros (value) {
+  value = value.toString()
+  if (value.indexOf('.') === -1) {
+    return value
+  }
+
+  let cutFrom = value.length - 1
+  do {
+    if (value[cutFrom] === '0') {
+      cutFrom--
+    }
+  } while (value[cutFrom] === '0')
+
+  if (value[cutFrom] === '.') {
+    cutFrom--
+  }
+
+  return value.substr(0, cutFrom + 1)
+}
+
+let Dygraph // lazy loaded on connect
+var height, dcrPrice, hashrate, tpSize, tpValue, tpPrice, graphData, currentPoint
+
+function rateCalculation (y) {
+  y = y || 0.99 // 0.99 TODO confirm why 0.99 is used as default instead of 1
+  let x = 1 - y
+
+  // equation to determine hashpower requirement based on percentage of live stake
+  // https://medium.com/decred/decreds-hybrid-protocol-a-superior-deterrent-to-majority-attacks-9421bf486292
+  // (6 (1-f_s)⁵ -15(1-f_s) + 10(1-f_s)³) / (6f_s⁵-15f_s⁴ + 10f_s³)
+  // (6x⁵-15x⁴ +10x³) / (6y⁵-15y⁴ +10y³) where y = this.value and x = 1-y
+  return ((6 * Math.pow(x, 5)) - (15 * Math.pow(x, 4)) + (10 * Math.pow(x, 3))) / ((6 * Math.pow(y, 5)) - (15 * Math.pow(y, 4)) + (10 * Math.pow(y, 3)))
+}
+
+const deviceList = {
+  '0': {
+    hashrate: 34, // Th/s
+    units: 'Th/s',
+    power: 1610, // W
+    cost: 1282, // $
+    name: 'DCR5',
+    link: 'https://www.cryptocompare.com/mining/bitmain/antminer-dr5-blake256r14-34ths/'
+  },
+  '1': {
+    hashrate: 44, // Th/s
+    units: 'Th/s',
+    power: 2200, // W
+    cost: 4199, // $
+    name: 'D1',
+    link: 'https://www.cryptocompare.com/mining/crypto-drilling/microbt-whatsminer-d1-plus-psu-dcr-44ths/'
+  }
+}
+
+function legendFormatter (data) {
+  var html = ''
+  if (data.x == null) {
+    let dashLabels = data.series.reduce((nodes, series) => {
```

Comment:
If `data.x == null`, then there's actually nothing to display, so just `return ''` and the box won't even be drawn, if I'm remembering correctly.

Reply:
The idea is to show `N/A` instead of an empty space for such situations

Reply:
Why, though? To me, when it just randomly says N/A with no context, I assume an error has occurred.